Adding parser for the Thurman 2012 dataset.

DSuveges opened the pull request:
Similar in design to the other datasets, however this one is way simpler: 
- [x] Parsing input file following a pre-defined schema. + dropping some columns
- [x] Mapping genomic coordinates to the new build.
- [x] Joining with gene index based on gene symbol to get target identifier.
- [x] Stand alone script, that parsers and saves the output.
- [x] Integrated into the intervals pipeline.

DSuveges committed:
Update on config, andersson parser.

DSuveges committed:
Adding notebook to pre-process the Javierre dataset.

DSuveges committed:
Adding javierre parser.

DSuveges committed:
working draft of the javierre parser.

DSuveges committed:
Optimizing the join with the genes.

DSuveges committed:
finalizing Javierre parser.

DSuveges committed:
first draft of the jung parser.

DSuveges committed:
Finalizing jung2019 parser.

DSuveges committed:
First version of the Thurman parser.

DSuveges committed:
One more round of syncronizing the parsers for a more consistent look.

DSuveges committed:
fixing config

DSuveges commented:
On the ds-single dataproc cluster (16 cores) the parser runs for all the datasets in 6 minutes:
```bash
python intervals_parser.py
```

```
intervals_parser.py:20: UserWarning:
The version_base parameter is not specified.
Please specify a compatability version level, or None.
Will assume defaults for version 1.1
  @hydra.main(config_path=f'{script_folder}/configs', config_name='config')
/home/dsuveges/.local/lib/python3.8/site-packages/hydra/_internal/hydra.py:119: UserWarning: Future Hydra versions will no longer change working directory at job runtime by default.
See https://hydra.cc/docs/next/upgrades/1.1_to_1.2/changes_to_job_working_dir/ for more information.
  ret = run_job(
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/07/27 22:21:37 INFO org.apache.spark.SparkEnv: Registering MapOutputTracker
22/07/27 22:21:37 INFO org.apache.spark.SparkEnv: Registering BlockManagerMaster
22/07/27 22:21:37 INFO org.apache.spark.SparkEnv: Registering BlockManagerMasterHeartbeat
22/07/27 22:21:37 INFO org.apache.spark.SparkEnv: Registering OutputCommitCoordinator
[2022-07-27 22:21:47,486][root][INFO] - Parsing Andersson 2014 data...
[2022-07-27 22:21:54,713][root][INFO] - Parsing Javierre 2016 data...
[2022-07-27 22:21:56,839][root][INFO] - Parsing Jung 2019 data...
[2022-07-27 22:21:58,636][root][INFO] - Parsing Thurman 2016 data...
22/07/27 22:23:14 WARN org.apache.spark.storage.BlockManager: Block rdd_264_3 already exists on this machine; not re-adding it
22/07/27 22:26:29 WARN org.apache.spark.storage.BlockManager: Block rdd_229_186 already exists on this machine; not re-adding it
```

Output parquet to take a look: `gs://genetics-portal-dev-data/dev_intervals/interval_220727/`

In the upcoming updates, I'll sort out the proper integration of the pipeline into a single dataproc job, describing the requirements to build the environment + take a deeper look into how to resolve tissues and cell types when present. (There's a ticket around this "bio-feature hack", that needs our attention)

DSuveges committed:
Fixing bug in Javierre parser

DSuveges committed:
Fixing bug about liftover parameters

DSuveges committed:
Fixing bug in liftover, so filter is more robust.

DSuveges committed:
fixing bug.

ireneisdoomed commented on configs/config.yaml:
```diff
@@ -15,38 +15,20 @@ max_len_dff: 100 # Max difference in length between old and new interval (in bas
 
 intervals:
   liftover_chain_file: gs://genetics-portal-input/v2g_input/grch37_to_grch38.over.chain
+  gene_index: gs://genetics-portal-input/v2g_input/geneset.protein_coding.parquet
```
How is this dataset produced?

DSuveges replied:
The gene set is an interim ad hoc thing at the moment. I'm expecting the inclusion of the TSS and the canonical transcript annotation in the gene index. Once it is in place, this part will be solidified.

ireneisdoomed commented on intervals_parser.py:
```diff
@@ -18,31 +20,55 @@
+    (
+        df
+        # . The dataset needs to be repartitioned
+        .write.mode('overwrite').parquet(cfg.intervals.output + f'/interval_{version}')
```
Did you mean to repartition the data here?

DSuveges replied:
Yes, currently there are just too many partitions. I wanted to consolidate this by ordering by chromosome/position and partition by chromosome. Thanks for reminding me, that it is not yet there.

ireneisdoomed commented on modules/andersson2014_parser.py:
```diff
@@ -36,8 +36,11 @@ class parse_anderson:
     EXPERIMENT_TYPE = 'fantom5'
     PMID = '24670763'
     BIO_FEATURE = 'aggregate'
+    TWOSIDED_THRESHOLD = 2.45e6  # <-  this needs to phased out. Filter by percentile instead of absolute value.
```
Do you mean that this parameter should be calculated dynamically?

DSuveges replied:
Exactly, but at this point it is very hard to justify investing the effort to implement the logic.

ireneisdoomed commented on modules/andersson2014_parser.py:
```diff
@@ -88,28 +91,30 @@ def __init__(self, anderson_data_file: str, gene_index: dataframe, lift: LiftOve
         )
 
-    def get_anderson_intervals(self) -> dataframe:
+    def get_intervals(self) -> dataframe:
```
Why would you leave the logic of this function in the init instead of putting it inside `get_intervals`?

DSuveges replied:
Yes. That's right. I though I have consolidated the methods names. I'll do it.

ireneisdoomed commented on modules/andersson2014_parser.py:
```diff
@@ -36,8 +36,11 @@ class parse_anderson:
+        logging.info('Parsing Andersson 2014 data...')
 
         # Read the anderson file:
         parserd_anderson_df = (
```
I'd create a `parse_anderson` method.

ireneisdoomed commented on modules/andersson2014_parser.py:
```diff
@@ -171,13 +175,11 @@ def main(anderson_data_file: str, gene_index_file: str, chain_file: str, proximi
     logging.info(f'Adnerson file: {args.anderson_file}')
```
```suggestion
    logging.info(f'Anderson file: {args.anderson_file}')
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+
+    @staticmethod
+    def prepare_gene_index(gene_index: dataframe) -> dataframe:
+        '''Pre-processind the gene dataset
```
```suggestion
        '''Pre-processing the gene dataset
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+        self.javierre_intervals.write.mode('overwrite').parquet(output_file)
+
+    @staticmethod
+    def prepare_gene_index(gene_index: dataframe) -> dataframe:
```
What do you think if, instead of passing the df as an argument, you also delegate the function of reading the data here? This also applies to the other parsers.

DSuveges replied:
I'm not sure about that.... Let's wait till we have the target index with the required fields, the we need to consolidate this piece of logic anyway.

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+    spark = (
+        pyspark.sql.SparkSession.builder.config(conf=spark_conf)
+        .master('local[*]')
+        .config("spark.driver.bindAddress", "127.0.0.1")
```
What is this config for? If it is necessary we could move it to `spark_conf`

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+        javierre_raw = (
+            SparkSession.getActiveSession().read.parquet(javierre_parquet)
+
+            # SPlitting name column into chromosome, start, end, and score:
```
```suggestion
            # Splitting name column into chromosome, start, end, and score:
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser('Wrapper for the the Javierre interval data parser.')
+    parser.add_argument('--javierre_file', type=str, help='Path to the pre-processed parquet dataset.')
```
```suggestion
    parser.add_argument('--javierre_file', type=str, help='Path to the pre-processed parquet dataset (.parquet).')
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+        # Lifting over intervals:
+        javierre_remapped = (
+            javierre_raw
+            # Lifting over to GRCh39 interval 1:
```
```suggestion
            # Lifting over to GRCh38 interval 1:
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+            .withColumnRenamed('mapped_start', 'start')
+            .withColumnRenamed('mapped_end', 'end')
+
+            # Lifting over interval 2 to GRCh39:
```
```suggestion
            # Lifting over interval 2 to GRCh38:
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+        logging.info('Parsing Jung 2019 data...')
+
+        # Read Javierre data:
```
```suggestion
        # Read Jung data:
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+        self.jung_intervals = (
+            jung_raw
+
+            # Lifting over to GRCh39 interval 1:
```
```suggestion
            # Lifting over to GRCh38 interval 1:
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+    def qc_intervals(self) -> None:
+        """
+        Perform QC on the anderson intervals.
```
```suggestion
        Perform QC on the Jung intervals.
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+    spark = (
+        pyspark.sql.SparkSession.builder.config(conf=spark_conf)
+        .master('local[*]')
+        .config("spark.driver.bindAddress", "127.0.0.1")
```
Same comment as above

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+    # Initialze the parser:
+    logging.info('Starting Jung 2019 data processing.')
+    javierre = parse_jung(jung_data_file, gene_index, lift)
```
```suggestion
    jung = parse_jung(jung_data_file, gene_index, lift)
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+    # run QC:
+    logging.info('Running QC on the intervals.')
+    javierre.qc_intervals()
```
```suggestion
    jung.qc_intervals()
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+    # Save data:
+    logging.info(f'Saving data to {output_file}.')
+    javierre.save_parquet(output_file)
```
```suggestion
    jung.save_parquet(output_file)
```

ireneisdoomed commented on modules/jung2019_parser.py:
```diff
@@ -0,0 +1,171 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser('Wrapper for the the Jung interval data parser.')
+    parser.add_argument('--jung_file', type=str, help='Path to the raw csv dataset.')
```
```suggestion
    parser.add_argument('--jung_file', type=str, help='Path to the raw csv dataset (.csv).')
```

ireneisdoomed commented on modules/andersson2014_parser.py:
```diff
@@ -123,12 +128,12 @@ def save_parquet(self, output_file: str) -> None:
         pyspark.sql.SparkSession
         .builder
-        .master("local[*]")
+        # .master("local[*]")
```
Are you commenting this out for a reason?

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+    """
+    Parser Thurman 2012 dataset
+
+    :param Thurman_parquet: path to the parquet file containing the Thurman 2016 data
```
```suggestion
    :param Thurman_parquet: path to the parquet file containing the Thurman 2012 data
```

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+    :param gene_index: Pyspark dataframe containing the gene index
+    :param lift: LiftOverSpark object
+
+    **Summary of the logic:**
```
This is missing

DSuveges replied:
Ah, I'll add that one.

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+                 gene_index: dataframe,
+                 lift: LiftOverSpark) -> None:
+
+        logging.info('Parsing Thurman 2016 data...')
```
```suggestion
        logging.info('Parsing Thurman 2012 data...')
```

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+    spark = (
+        pyspark.sql.SparkSession.builder.config(conf=spark_conf)
+        .master('local[*]')
+        .config("spark.driver.bindAddress", "127.0.0.1")
```
Same comment as above

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+        .getOrCreate()
+    )
+
+    logging.info('Reading genes and initializeing liftover.')
```
```suggestion
    logging.info('Reading genes and initializing liftover.')
```

ireneisdoomed commented on modules/thurman2012_parser.py:
```diff
@@ -0,0 +1,165 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser('Wrapper for the the Thurman interval data parser.')
+    parser.add_argument('--thurman_file', type=str, help='Path to the tsv dataset.')
```
```suggestion
    parser.add_argument('--thurman_file', type=str, help='Path to the tsv dataset (.tsv).')
```

ireneisdoomed commented on modules/javierre2016_parser.py:
```diff
@@ -0,0 +1,238 @@
+    """
+    Parser Javierre 2016 dataset
+
+    :param javierre_parquet: path to the parquet file containing the Javierre 2016 data
```
```suggestion
    :param javierre_parquet: path to the parquet file containing the Javierre 2016 data after processing it (see notebooks/Javierre_data_pre-process.ipynb)
```

ireneisdoomed commented on configs/config.yaml:
```diff
@@ -15,38 +15,20 @@ max_len_dff: 100 # Max difference in length between old and new interval (in bas
-    pubmed_id: '27863249'
-    data_type: interval
-    experiment_type: pchic
+  javierre_dataset: gs://genetics-portal-input/v2g_input/javierre_2016_preprocessed.parquet
```
I'd change the filename to `javierre_file` for consistency

DSuveges replied:
It's not really a file *per se*, a dataset is partitioned and saved as a series parquet files.

ireneisdoomed reviewed (commented):
Awesome job. I've submitted a lot of comments which are mostly related to style. The logic is clear.

Overall my only main reservation is the use of the classes here (I'm not a big fan of them overall). I really like how the liftover class is instantiated, e.g. But I'm not sure I see the point in this case when we are only instantiating them once. For me, classes are for modularisation, and if you see the code of each class for each type of dataset, there are 3 functions that are being repeated constantly:
- Lifting the coordinates over
- Logging some stats (I really like this)
- Saving to parquet

In my opinion it would be nice to extract these common methods somewhere (single functions or a common `parse_interval_utils` class so that the code for each dataset is slimmer.

DSuveges commented:
> Awesome job. I've submitted a lot of comments which are mostly related to style. The logic is clear.
> 
> Overall my only main reservation is the use of the classes here (I'm not a big fan of them overall). I really like how the liftover class is instantiated, e.g. But I'm not sure I see the point in this case when we are only instantiating them once. For me, classes are for modularisation, and if you see the code of each class for each type of dataset, there are 3 functions that are being repeated constantly:
> 
> * Lifting the coordinates over
> * Logging some stats (I really like this)
> * Saving to parquet
> 
> In my opinion it would be nice to extract these common methods somewhere (single functions or a common `parse_interval_utils` class so that the code for each dataset is slimmer.

The question of the code orgranization is up to debate, for me it looks like a question of taste, given it more or less works, I would not invest into restructure.

DSuveges committed:
Apply suggestions from code review

Co-authored-by: Irene López <45119610+ireneisdoomed@users.noreply.github.com>

DSuveges committed:
Addressing general comment on PR

DSuveges committed:
Slight changes.

DSuveges committed:
adding extra logging.

DSuveges committed:
bugfix

DSuveges committed:
Fixing bug in liftover

DSuveges committed:
Fixing some logging.

DSuveges commented:
I have addressed most of the issues that were raised in the review. I also updated the Liftover script, as the old version was not 1 based when lifting over intervals. There's added logging to have a better idea on how the process is progressing. The process takes ~6minutes and output is ordered by chromosome and position and repartitioned into 200 partitions.

DSuveges commented:
This repo no longer developed.
